### ISSUE TITLE
Using the automation API with SSH URLs does not utilize ambient credentials

### ISSUE THREAD

**Issue by AaronFriel** [pulumi/pulumi#16965]
### What happened?

Using the Pulumi Automation API `NewLocalWorkspace` with a private repository as the URL throws an error.

This is "hard to use correctly" because the `GitAuth` configuration accepts an `SSHPrivateKeyPath`, but it must be a specific key. For software shared among many developers, that key may vary and not exist in a well known location, though SSH libraries are capable of enumerating and trying multiple keys in valid locations.

### Example

```go
w, err := auto.NewLocalWorkspace(ctx, auto.Repo(auto.GitRepo{
  URL: "...", // private SSH url
  // ...
}))
```

Error:
```
failed to create workspace, unable to enlist in git repo: unable to clone repo: error creating SSH agent: "SSH agent requested but SSH_AUTH_SOCK not-specified
```

### Output of `pulumi about`

```
CLI          
Version      3.129.1-alpha.1723578932
Go Version   go1.22.6
Go Compiler  gc

Host     
OS       ubuntu
Version  22.04
Arch     x86_64

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/friel-pulumi-corp
User           friel-pulumi-corp
Organizations  friel-pulumi-corp, friels-test-org, upstarts, pulumi
Token type     personal

Pulumi locates its logs in /tmp by default
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by AaronFriel** [comment]
Adding some background research on this and why we see this.

My current assumption is:
1. We are falling back to SSH agent too quickly, having failed to find an identity file.

Red herring: Per a number of open issues in go-git, the library does not fully support Git Protocol v2. **Is it relevant?** I cannot find anything that suggests protocol v2 is required to utilize multiple SSH keys.
* https://github.com/go-git/go-git/issues/64

Likely avenue of success:

The default `IdentityFile` path returned from `sshConfig.GetStrict` during debugging is `~/.ssh/identity`, which is typically not populated, a legacy of older SSH clients.
https://github.com/pulumi/pulumi/blob/a3d4459b824b5b6dd72f424dfddbc9c957bb76e0/sdk/go/common/util/gitutil/git.go#L368

Go-Git's `NewPublicKeysFromFile` and other helper methods don't allow specifying multiple files, either:
https://github.com/go-git/go-git/blob/302dddeda962e4bb3477a8e4080bc6f5a253e2bb/plumbing/transport/ssh/auth_method.go#L132-L157

But as implied by the name, the `ssh.PublicKeys` constructor is plural, and uses a variadic argument to accept multiple signers. We can construct our own.

We should be able to:

1. Add a struct specifying multiple files for `ssh.PublicKeys`
2. Add a struct for muxing auth and implementing `ssh.PublicKeysCallback`, combining the result of (1.) and if an SSH agent is available, its return value of `ssh.PublicKeysCallback`
3. Supply a proper list of all valid identity files to (1.)

**Comment by tgummerer** [comment]
> Red herring: Per a number of open issues in go-git, the library does not fully support Git Protocol v2. Is it relevant? I cannot find anything that suggests protocol v2 is required to utilize multiple SSH keys.

No, the Git protocol version has nothing to do with SSH setup. SSH setup happens first and only then Git speaks its own protocol.
